### DISCUSSION TITLE
Filtering based on IDictionary values

### DISCUSSION THREAD

**Discussion by ameckl** [alirezanet/Gridify#164] [Q&A]
> I'm currently using Sieve's custom filter methods to achieve this functionality, but planning to migrate to Gridify.

Hello, I'm using MongoDb as database, where I have a couple of entities with sub-documents storing additional information, represented as `IDictionary<string, string>`.

e.g.
```c#
public class ExampleEntity
{
   public IDictionary<string, string> Data { get; init; }
}
```

Is it possible to achieve filtering based on the values of this Dictionary?

> The keys of these dictionaries are not known, so I cannot just `AddMap("data.key1"), AddMap("data.key2")`.

One solution would be to implement a custom operator, then cast the `value` to an IDictionary, but that seems like an overkill (and then I need to implement other custom operators for `NotEquals, Contains`, etc).

It would be great to have support on sub-documents, just like there is support for [sub-collections](https://alirezanet.github.io/Gridify/guide/gridifyMapper#use-indexes-on-sub-collections). Maybe something like:

```c#
var gq = new GridifyQuery { Filter = "prop{subKey} = John"};

var gm = new GridifyMapper<TargetType>()
      .AddMap("prop", (x , prop) => x.Data[prop]);

// would generate an IQueryable like

IQueryable<TargetType>().Where(e => e.Data["subKey"] == "John");
```

For reference, my current approach (using Sieve) looks like this:

```c#
public IQueryable<ExampleEntity> DataValue(IQueryable<ExampleEntity> source, string op, string[] values)
    {
        if (values.Length != 2)
        {
            return source;
        }

        string key = values.First();
        string value = values.Last();

        if (op == "==")
        {
            source = source.Where(d => d.Data[key] == value);
        }
        // ... the rest of operators   

        return source;
    }

// where
IQueryable<ExampleEntity> source = MongoCollection.AsQueryable();
// then GET /api/example?filters="DataValue==key1|value1"
```

Or maybe is it planned to support custom filter methods (like Sieve), where we have access to the `IQueryable` that is being filtered? (that's how we achieved search in our past projects) 

e.g.

```c#
// really basic example
public IQueryable<ExampleEntity> Search(IQueryable<ExampleEntity> source, string op, string[] values)
    {
        
        string value = values.First();
        // here we have the possibility to access the Queryable and to call database engine specific clauses like a Mongo text search.
        source = source.Where(d => d.Name.Contains(value) || d.Description.Contains(value))

        return source;
    }

// then GET /api/example?filters="search==searchString"
```

**Comment by alirezanet** [comment]
Hi @ameckl,
Thanks for the feedback

> Is it possible to achieve filtering based on the values of this Dictionary?

I'll take a look at this maybe in future versions we could add Dictionary built-in support , it would be a cool additional feature 👌

---

> Or maybe is it planned to support custom filter methods (like Sieve), where we have access to the IQueryable that is being filtered? (that's how we achieved search in our past projects)

You already **have** access to `IQueryable` if you want to combine it with GridifyQuery manually. 

```csharp
var gq = new GridifyQuery { Filter = "Name = John"};

var query = DS
	.Where(ds =>  ds.Data["key"] == "value")
	.ApplyFiltering(gq);
```
or do you need to access the value and operator?

---

**Reply by alirezanet** [reply]
by the way, I created an issue here #165 to take a look further.

**Comment by alirezanet** [comment]
Hello @ameckl,
Could you give version `2.15.0-preview1` a try? This version of Gridify now supports all types of indexers, which means you can even apply it to dictionaries. Here's an example:

```csharp
var dataSource = new List<Test>()
{
	new() { Id = 1, MyDictionary = new() { { "Key1", "Value1" }, { "Key2", "Value2" } } },
	new() { Id = 2, MyDictionary = new() { { "Key3", "Value1" }, { "Key4", "Value2" } } },
	new() { Id = 3, MyDictionary = new() { { "Key5", "Value3" }, { "Key6", "Value4" } } },
}.AsQueryable();


var qb = new QueryBuilder<Test>()
	.AddMap("prop", (target, key) => target.MyDictionary[key])
	.AddCondition("prop[Key4] = Value2");

qb.Build(dataSource).Select(q => q.Id).Dump(); // returns Id 2


class Test
{
	public int Id { get; set; }
	public Dictionary<string,string> MyDictionary{ get; set; }
}
```

I'm currently evaluating this feature, but in the meantime, your feedback would be greatly appreciated.

**Reply by ameckl** [reply]
Hi @alirezanet!

It really looks like an interesting and useful improvement. 

In the meantime we switched our database to PostgreSQL and I moved all my `IDictionary<string,string>`s to `JsonElement` so I can use `jsonb[]` column type. But anyways, there should be no big difference between them.

I'll give it a try (hopefully) next week.

**Reply by moxplod** [reply]
Did you ever try this?
